apps/apache.go

```diff
@@ -0,0 +1,171 @@
+		// Sample "combined" line: ::1 - - [26/Aug/2021:16:49:43 +0000] "GET / HTTP/1.1" 200 10701 "-" "curl/7.64.0"
+		Regex: `^(?<http_request_remoteIp>[^ ]*) (?<host>[^ ]*) (?<user>[^ ]*) \[(?<time>[^\]]*)\] "(?<http_request_requestMethod>\S+)(?: +(?<http_request_requestUrl>[^\"]*?)(?: +(?<http_request_protocol>\S+))?)?" (?<http_request_status>[^ ]*) (?<http_request_responseSize>[^ ]*)(?: "(?<http_request_referer>[^\"]*)" "(?<http_request_userAgent>[^\"]*)")?$`,
+		LoggingProcessorParseShared: confgenerator.LoggingProcessorParseShared{
+			TimeKey:    "time",
+			TimeFormat: "%d/%b/%Y:%H:%M:%S %z",
+			Types: map[string]string{
+				"http_request_status": "integer",
+				// N.B. "http_request_responseSize" is a string containing an integer.
+				// https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#HttpRequest.FIELDS.response_size
+			},
+		},
+	}.Components(tag, uid)
+	// apache logs "-" when a field does not have a value. Remove the field entirely when this happens.
+	for _, field := range []string{
+		"host",
+		"user",
+		"http_request_referer",
+	} {
+		c = append(c, fluentbit.Component{
+			Kind: "FILTER",
+			Config: map[string]string{
+				"Name":      "modify",
+				"Match":     tag,
+				"Condition": fmt.Sprintf("Key_Value_Equals %s -", field),
+				"Remove":    field,
+			},
+		})
+	}
+	// Generate the httpRequest structure.
+	c = append(c, fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Name":          "nest",
+			"Match":         tag,
+			"Operation":     "nest",
+			"Wildcard":      "http_request_*",
+			"Nest_under":    "logging.googleapis.com/http_request",
+			"Remove_prefix": "http_request_",
+		},
+	})
+	return c
+}
+
+type LoggingProcessorApacheError struct {
+	confgenerator.ConfigComponent `yaml:",inline"`
+}
+
+func (LoggingProcessorApacheError) Type() string {
+	return "apache_error"
+}
+
+func (p LoggingProcessorApacheError) Components(tag string, uid string) []fluentbit.Component {
+	c := confgenerator.LoggingProcessorParseRegex{
+		// Documentation: https://httpd.apache.org/docs/current/logs.html#errorlog
+		// Sample line 2.4: [Fri Sep 09 10:42:29.902022 2011] [core:error] [pid 35708:tid 4328636416] (13)Permission denied [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
+		// Sample line 2.2: [Fri Sep 09 10:42:29.902022 2011] [error] [pid 35708:tid 4328636416] [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
+		// TODO - Support time parsing for version 2.0 where smallest resolution is seconds
+		// Sample line 2.0: [Wed Oct 11 14:32:52 2000] [error] [client 127.0.0.1] client denied by server configuration: /export/home/live/ap/htdocs/test
+		Regex: `^\[(?<time>[^\]]+)\] \[(?:(?<module>\w+):)?(?<level>[\w\d]+)\](?: \[pid (?<pid>\d+)(?::tid (?<tid>[0-9]+))?\])?(?: (?<errorCode>[^\[]*))?(?: \[client (?<client>[^\]]*)\])? (?<message>.*)$`,
```

Comment:
Will this pattern cause the message to end up in the errorCode field when the error log does not contain a `[client` field? e.g. here's a real error log entry from 2.4:

```
[Thu Sep 30 03:18:29.239182 2021] [ssl:error] [pid 2451:tid 140169666050176] AH02217: ssl_stapling_init_cert: Can't retrieve issuer certificate!
```

Reply:
Yeah, this fails to parse that message appropriately. The below would capture the message you sent properly, but if we're missing a semicolon at the end of the errorCode it would fail. Do you think we should try to capture error code even if it could be fragile, or just capture errorCode only if it precedes a client and is clear like that.
```
^\[(?<time>[^\]]+)\] \[(?:(?<module>\w+):)?(?<level>[\w\d]+)\](?: \[pid (?<pid>\d+)(?::tid (?<tid>[0-9]+))?\])?(?: (?<errorCode>[^\[:]*):?)?(?: \[client (?<client>[^\]]*)\])? (?<message>.*)$
```

Reply:
I'd lean towards something more restricted and we can loosen it if needed because it's forward compatible. Because the other way around could be considered a breaking change.

Reply:
@qingling128 my understanding is that you mean to use the above new regex, in order to capture errorCode when available, even thought it could be fragile. Is that correct?

Reply:
It seems like the new regex above is not able to capture the 3 log formats in the comment:

```
		// Sample line 2.4: [Fri Sep 09 10:42:29.902022 2011] [core:error] [pid 35708:tid 4328636416] (13)Permission denied [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
		// Sample line 2.2: [Fri Sep 09 10:42:29.902022 2011] [error] [pid 35708:tid 4328636416] [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
		// Sample line 2.0: [Wed Oct 11 14:32:52 2000] [error] [client 127.0.0.1] client denied by server configuration: /export/home/live/ap/htdocs/test
```

Reply:
It works properly from my understanding, and I just went and validated with regex101.com as well. What fields are you concerned about?

Reply:
Maybe I missed something, regex101 did not parse those 3 when I tried. For instance:

![image](https://user-images.githubusercontent.com/5287526/135899102-02080be6-305f-4bf2-9f0f-fcefe5da764e.png)

Reply:
Sync'd offline. It was the newline that accidentally failed the match. The proposed new regex actually work for all 4 cases.

Reply:
The latest change LGTM.